### PR TITLE
chore: update plugins, typescript, prettier deps

### PR DESCRIPTION
**Pull request by ndhoule**
This changeset updates all `eslint-plugin-` dependencies and any of
their peers (i.e. `prettier` and `typescript`).

This enables a couple new rules, and some of the plugins themselves had
breaking changes (a couple rule renames, but mostly changes to their
recommended rulesets) so this is a breaking change.

### CODE REVIEW

**Comment by ndhoule** [comment]
!merge

**Comment by ndhoule** [comment]
!merge

File: package.json [hunk 1]
```diff
@@ -25,47 +25,47 @@
...
-    "prettier": "1.18.2",
-    "typescript": "*"
+    "eslint-plugin-jest": "^23.0.3",
+    "eslint-plugin-lodash": "^6.0.0",
+    "eslint-plugin-mocha": "^6.2.1",
+    "eslint-plugin-prettier": "^3.1.1",
+    "eslint-plugin-react": "^7.16.0",
+    "eslint-plugin-react-hooks": "^2.2.0",
+    "prettier": ">=1.13.0",
+    "typescript": ">=3.2.1 <3.8.0"
```
Comment (serhalp) [inline package.json]: any context for this one? why cap it at 3.7.x?

File: src/config/jest-strict.js [hunk 2]
```diff
@@ -11,11 +11,9 @@ export default {
   plugins: ['goodeggs', 'jest'],
   extends: ['plugin:goodeggs/jest', 'plugin:jest/style'],
   rules: {
-    'jest/expect-expect': 'warn',
     'jest/lowercase-name': 'error',
-    'jest/no-alias-methods': 'error',
     'jest/no-truthy-falsy': 'error',
-    'jest/require-tothrow-message': 'error',
+    'jest/require-to-throw-message': 'error',
```
Comment (serhalp) [inline src/config/jest-strict.js]: lol wat

File: src/config/mocha.js [hunk 3]
```diff
@@ -3,11 +3,18 @@ export default {
...
   extends: ['plugin:mocha/recommended'],
   rules: {
-    'mocha/handle-done-callback': 'error',
-    'mocha/no-async-describe': 'error',
-    'mocha/no-identical-title': 'error',
-    'mocha/no-nested-tests': 'error',
-    'mocha/no-return-and-callback': 'error',
+    // TODO(ndhoule): Enabled in `mocha/recommended`. This is a de facto best practice, but we
+    // haven't codified it anywhere; consider enabling this rule.
+    'mocha/max-top-level-suites': 'off',
```
Comment (serhalp) [inline src/config/mocha.js]: oh hell yes, this rule exists?

File: src/config/typescript.js [hunk 4]
```diff
@@ -13,15 +13,21 @@ export default {
...
     '@typescript-eslint/indent': 'off',
+    // TODO(ndhoule): Enabled by @typescript-eslint/recommended. Discuss whether or not this is
+    // worthwhile; I've never had it cause any pain, but perhaps others have?
+    '@typescript-eslint/no-empty-function': 'off',
     // We often use empty interfaces for e.g. props factories for React components that don't yet,
     // but will eventually, accept any props.
     '@typescript-eslint/no-empty-interface': 'off',
     '@typescript-eslint/no-for-in-array': 'error',
+    // TODO(ndhoule): Enabled by @typescript-eslint/recommended. Discuss enabling it permanently.
+    '@typescript-eslint/no-this-alias': 'off',
```
Comment (serhalp) [inline src/config/typescript.js]: Not sure how relevant this is, but we have a sort of best practice within a bad practice, which is to always alias `this` in legacy mongoose methods/statics to make the code more readable. (e.g. `const pickInstruction = this;` at the start of the method)

File: package.json (hunk 1 above)
Comment (ndhoule) [reply package.json]: I wish I could leave comments in `package.json`. This is the range specified in the `@typescript-eslint/parser` documentation.

File: src/config/jest-strict.js (hunk 2 above)
Comment (ndhoule) [reply src/config/jest-strict.js]: Rule got renamed

File: src/config/mocha.js (hunk 3 above)
Comment (ndhoule) [reply src/config/mocha.js]: Yes! We should enable it sometime. (It _destroyed_ Garbanzo.)

File: src/config/typescript.js (hunk 4 above)
Comment (ndhoule) [reply src/config/typescript.js]: Yeah, I'm not totally convinced this is helpful; it's certainly not essential, and we violate it a lot. (But it's true that arrow functions and proper scoping mostly eliminate the need for it.)

**Pull request merged**